CLI/digest.c

```diff
@@ -0,0 +1,166 @@
+        int error_code = psabpf_digest_open(psabpf_ctx, ctx, **argv);
+        if (error_code != NO_ERROR) {
+            fprintf(stderr, "failed to open digest %s: %s\n", **argv, strerror(error_code));
+            return error_code;
+        }
+        *instance_name = **argv;
+    }
+    NEXT_ARGP();
+
+    return NO_ERROR;
+}
+
+static int build_struct_json(json_t *parent, psabpf_digest_context_t *ctx, psabpf_digest_t *digest)
+{
+    psabpf_digest_field_t *field;
+    while ((field = psabpf_digest_get_next_field(ctx, digest)) != NULL) {
+        /* To build flat structure of output JSON just remove this and next conditional
+         * statement. In other words, preserve only condition and instructions below it:
+         *      if (psabpf_digest_get_field_type(field) != DIGEST_FIELD_TYPE_DATA) continue; */
+        if (psabpf_digest_get_field_type(field) == DIGEST_FIELD_TYPE_STRUCT_START) {
+            json_t *sub_struct = json_object();
+            if (sub_struct == NULL) {
+                fprintf(stderr, "failed to prepare message sub-object JSON\n");
+                return ENOMEM;
+            }
+            if (json_object_set(parent, psabpf_digest_get_field_name(field), sub_struct)) {
+                fprintf(stderr, "failed to add message sub-object JSON\n");
+                json_decref(sub_struct);
+                return EPERM;
+            }
+
+            int ret = build_struct_json(sub_struct, ctx, digest);
+            json_decref(sub_struct);
+            if (ret != NO_ERROR)
+                return ret;
+
+            continue;
+        }
+
+        if (psabpf_digest_get_field_type(field) == DIGEST_FIELD_TYPE_STRUCT_END)
+            return NO_ERROR;
+
+        if (psabpf_digest_get_field_type(field) != DIGEST_FIELD_TYPE_DATA)
+            continue;
+
+        const char *encoded_data = convert_bin_data_to_hexstr(psabpf_digest_get_field_data(field),
+                                                              psabpf_digest_get_field_data_len(field));
+        if (encoded_data == NULL) {
+            fprintf(stderr, "not enough memory\n");
+            return ENOMEM;
+        }
+        const char *field_name = psabpf_digest_get_field_name(field);
+        if (field_name == NULL)
+            field_name = "";
+        json_object_set_new(parent, field_name, json_string(encoded_data));
+        free((void *) encoded_data);
+    }
+
+    return NO_ERROR;
+}
+
+int do_digest_get(int argc, char **argv)
+{
+    psabpf_context_t psabpf_ctx;
+    psabpf_digest_context_t ctx;
+    int error_code = EPERM;
+    const char *digest_instance_name = NULL;
+
+    psabpf_context_init(&psabpf_ctx);
+    psabpf_digest_context_init(&ctx);
+
+    if (parse_pipeline_id(&argc, &argv, &psabpf_ctx) != NO_ERROR)
+        goto clean_up_psabpf;
+
+    if (parse_digest(&argc, &argv, &psabpf_ctx, &ctx, &digest_instance_name) != NO_ERROR)
+        goto clean_up_psabpf;
+
+    if (argc > 0) {
+        fprintf(stderr, "%s: unused argument\n", *argv);
```

Comment:
Maybe it would be better to show only a warning ?

Reply:
I think it is better to make error for unused arguments:
- In [other places in our code](https://github.com/P4-Research/psabpf/blob/master/CLI/table.c#L256-L259) is done in this way.
- It is clear message to user that he made a mistake (`psabpf-ctl digest get` can obtain data from single digest instance, but what if user wants read two different Digests?)